### ISSUE TITLE
Rework zarr/dask chunking

### ISSUE THREAD

**Issue by amarcozzi** [silvxlabs/quicfire-tools#32]

**Comment by amarcozzi** [comment]
@zacharycope0 what do you think of the following options for reworking custom chunk options in the outputs module?

## Option 1:
Users can pass in a specified chunk size for a given output.
```python
# to_zarr method becomes
    def to_zarr(self, fpath: Path, chunks: dict = None):
        """Write the data to a zarr file."""
        # Create the zarr file
        zarr_file = zarr.open(str(fpath), mode="w")

        # Write each output to the zarr file
        for output_name, output in self.outputs.items():
            # Create a zarr dataset for the output
            shape = (len(output.times), *output.shape)
            if output_name in chunks:
                chunks = chunks[output_name]
            else:
                chunks = [1 if i == 0 else shape[i] for i in range(len(shape))]
            zarr_file.create_dataset(
                output_name,
                shape=shape,
                chunks=chunks,
                dtype=float)

            # Write each timestep to the output's zarr dataset
            for time_step in range(len(output.times)):
                data = self.to_numpy(output_name, time_step)
                zarr_file[output_name][time_step, ...] = data

        return zarr_file

# Then call as
sim_outputs = SimulationOutputs("Outputs")
my_chunks = {"surfaceEnergy": (10, 500, 500)}
sim_outputs.to_zarr("zarr_file", chunks=my_chunks)
```

## Option 2:
Make all arrays 4D instead of some arrays 3D and some arrays 4D. Then you could pass a chunk size in for all arrays if you give to_zarr a tuple, or specific arrays if you give it a dictionary. 

```python
# to_zarr becomes
    def to_zarr(self, fpath: Path, chunks: tuple[int] | dict[str: tuple[int]] = None):
        """Write the data to a zarr file."""
        # Create the zarr file
        zarr_file = zarr.open(str(fpath), mode="w")

        # Write each output to the zarr file
        for output_name, output in self.outputs.items():
            # Create a zarr dataset for the output
            shape = (len(output.times), *output.shape)
            if isinstance(chunks, dict) and output_name in chunks:
                chunks = chunks[output_name]
            elif isinstance(chunks, tuple):
                chunks = chunks
            else:
                chunks = [1 if i == 0 else shape[i] for i in range(len(shape))]
            zarr_file.create_dataset(
                output_name,
                shape=shape,
                chunks=chunks,
                dtype=float)

            # Write each timestep to the output's zarr dataset
            for time_step in range(len(output.times)):
                data = self.to_numpy(output_name, time_step)
                zarr_file[output_name][time_step, ...] = data

        return zarr_file

# Then call as
sim_outputs = SimulationOutputs("Outputs")
my_chunks = (10, 1, 500, 500)
# or my_chunks = {"surfaceEnergy": (10, 500, 500)}
sim_outputs.to_zarr("zarr_file", chunks=my_chunks)
```

## Option 3: Keep default chunking (maybe with room for an optional chunking parameter) and have users rechunk if needed.

```python
import zarr
from rechunker import rechunk

sim_outputs = SimulationOutputs("Outputs")
source = sim_outputs.to_zarr("zarr_file", chunks=my_chunks)

intermediate = "intermediate.zarr"
target = "target.zarr"
rechunked = rechunk(source, target_chunks=(4, 1), target_store=target,
                    max_mem=256000,
                    temp_store=intermediate)
rechunked.execute()
```


There are more than 3 options, but those are 3 decent ones that come to mind with various pros/cons. Let me know what you rhink and feel free to throw out other ideas/implementations. I agree that we should improve this functionality, though.

**Comment by amarcozzi** [comment]
As another option we can look into some more implementations and copy how they do it. I kind of like how dask handles "automatic chunking" here: https://docs.dask.org/en/latest/array-chunks.html#automatic-chunking.

There's some good information on that page that may help us make a better decision.

**Comment by zacharycope0** [comment]
@amarcozzi  Thanks for the suggestions. I was out for surgery, so I apologize for the late response. 

At this point, using the rechunker would make the most sense. The only output file I have had trouble working with is the SurfEng because the time dimension can get so large. How about if the user runs functions to analyze the surf_eng we have a function parameter that will rechunk the SurfEng zarr (rechunk_flag = True).

I'll run speed tests to make sure that rechunking to put the time dim in one chunk would be better than running the function with sub-optimal chunking.

**Comment by zacharycope0** [comment]
@amarcozzi Here is the issue I was describing to get zarrs loaded into xarray datasets with xr.open_zarr(). I just pushed a new commit to compare the two zarr hierarchies:
**Method 1 hierarchy: zarr_group <-  zarr_datasets (named with output file names)**
**Method 2 hierarchy: zarr_group <- zarr_groups (named with output file names) <-  zarr_dataset (named 'data')**

# Here are the two to_zarr() functions. Swap commented lines and delete the .zarr file to compare.
https://github.com/silvxlabs/quicfire-tools/blob/6aaa6f428614d5076e72eea8db377d056b4e8e6b/quicfire_tools/outputs.py#L458-L576

# Here is where I was testing
https://github.com/silvxlabs/quicfire-tools/blob/6aaa6f428614d5076e72eea8db377d056b4e8e6b/tests/test_LoadSurfEng_module.py#L57-L68

# to_zarr Errors Method 1: 
**ds = xr.open_zarr(simulation_outputs.get_output('surfEnergy').zarr_path)**
ds = xr.open_zarr(simulation_outputs.get_output('surfEnergy').zarr_path)
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/home/username/miniconda3/envs/ff_dask/lib/python3.11/site-packages/xarray/backends/zarr.py", line 867, in open_zarr
    ds = open_dataset(
         ^^^^^^^^^^^^^
  File "/home/username/miniconda3/envs/ff_dask/lib/python3.11/site-packages/xarray/backends/api.py", line 570, in open_dataset
    backend_ds = backend.open_dataset(
                 ^^^^^^^^^^^^^^^^^^^^^
  File "/home/username/miniconda3/envs/ff_dask/lib/python3.11/site-packages/xarray/backends/zarr.py", line 949, in open_dataset
    ds = store_entrypoint.open_dataset(
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/username/miniconda3/envs/ff_dask/lib/python3.11/site-packages/xarray/backends/store.py", line 43, in open_dataset
    vars, attrs = filename_or_obj.load()
                  ^^^^^^^^^^^^^^^^^^^^^^
  File "/home/username/miniconda3/envs/ff_dask/lib/python3.11/site-packages/xarray/backends/common.py", line 210, in load
    (_decode_variable_name(k), v) for k, v in self.get_variables().items()
                                              ^^^^^^^^^^^^^^^^^^^^
  File "/home/username/miniconda3/envs/ff_dask/lib/python3.11/site-packages/xarray/backends/zarr.py", line 513, in get_variables
    (k, self.open_store_variable(k, v)) for k, v in self.zarr_group.arrays()
                                                    ^^^^^^^^^^^^^^^^^^^^^^
AttributeError: 'Array' object has no attribute 'arrays'

**ds = xr.open_zarr(simulation_outputs.zarr_path)**
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/home/username/miniconda3/envs/ff_dask/lib/python3.11/site-packages/xarray/backends/zarr.py", line 867, in open_zarr
    ds = open_dataset(
         ^^^^^^^^^^^^^
  File "/home/username/miniconda3/envs/ff_dask/lib/python3.11/site-packages/xarray/backends/api.py", line 570, in open_dataset
    backend_ds = backend.open_dataset(
                 ^^^^^^^^^^^^^^^^^^^^^
  File "/home/username/miniconda3/envs/ff_dask/lib/python3.11/site-packages/xarray/backends/zarr.py", line 949, in open_dataset
    ds = store_entrypoint.open_dataset(
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/username/miniconda3/envs/ff_dask/lib/python3.11/site-packages/xarray/backends/store.py", line 58, in open_dataset
    ds = Dataset(vars, attrs=attrs)
         ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/username/miniconda3/envs/ff_dask/lib/python3.11/site-packages/xarray/core/dataset.py", line 652, in __init__
    variables, coord_names, dims, indexes, _ = merge_data_and_coords(
                                               ^^^^^^^^^^^^^^^^^^^^^^
  File "/home/username/miniconda3/envs/ff_dask/lib/python3.11/site-packages/xarray/core/merge.py", line 569, in merge_data_and_coords
    return merge_core(
           ^^^^^^^^^^^
  File "/home/username/miniconda3/envs/ff_dask/lib/python3.11/site-packages/xarray/core/merge.py", line 761, in merge_core
    dims = calculate_dimensions(variables)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/username/miniconda3/envs/ff_dask/lib/python3.11/site-packages/xarray/core/variable.py", line 3282, in calculate_dimensions
    raise ValueError(
ValueError: conflicting sizes for dimension 'time': length 1 on 'groundfuelheight' and length 7 on {'time': 'fire-energy_to_atmos', 'y': 'fire-energy_to_atmos', 'x': 'fire-energy_to_atmos', 'z': 'fire-energy_to_atmos'}

**Comment by amarcozzi** [comment]
Closed with #164
